core/src/upgrade/select.rs

```diff
@@ -70,11 +64,16 @@ where
     type Error = Either<EA, EB>;
     type Future = EitherFuture<A::Future, B::Future>;
 
-    fn upgrade_inbound(self, sock: C, info: Self::Info) -> Self::Future {
-        match info {
-            EitherName::A(info) => EitherFuture::First(self.0.upgrade_inbound(sock, info)),
-            EitherName::B(info) => EitherFuture::Second(self.1.upgrade_inbound(sock, info)),
+    fn upgrade_inbound(self, sock: C, selected_protocol: Protocol) -> Self::Future {
+        if self.0.protocols().any(|p| p == selected_protocol) {
```

Comment:
As described in your pull request, having to do this now is a bummer. This is in the hot-path. With our move to one-stream-per-request this will be executed often.

All that said, I am fine moving forward here. Based on intuition, this warrants a benchmark before introducing any optimization.

Reply:
One issue is that `SelectUpgrade` is constructed for each upgrade, meaning it is pointless to try and cache something.

Once we get rid of upgrades for protocols, we can try and cache some information in the handler, i.e. a HashMap of `Protocol` to sub-handler.

I'd suggest we defer it until then. Despite being the hotpath, I think iterating a handful of protocols (which should be allocation-free) shouldn't be very slow?

Reply:
An alternative that came to my mind:

Introduce a `ToProtocol` trait which `Protocol` implements by default. Our `ToProtocolsIter` would require an `Iterator` where `Item` implements `ToProtocol`.

With this abstraction we still have the simplicity for the user, namely that they can just use `Protocol` and that they can't mistake a `Protocol` for a `str` given that the latter is wrapped in a newtype. At the same time we maintain the functionality for e.g. the `SelectUpgrade`.

Just a thought. I don't feel strongly about this.

Reply:
That is effectively the same solution as we had before, just with a different name. It doesn't remove a "layer" for the user because they still have to specify the `Item` type.

But perhaps we should go with this path until we fully tackle: https://github.com/libp2p/rust-libp2p/issues/2863

At that point, we will also be able to implement performance optimizations in the `Select` handler.

Reply:
With the introduction of separate newtypes in `multistream-select` and `libp2p-core`, this is proving to be difficult.

On the one hand, I'd like to maintain this separation, I think `multistream-select` should be an implementation detail of `libp2p-core` and `libp2p-swarm`.

But, with the two new-types in place, I cannot re-create the type-safety we had before because `multistream-select` doesn't know about the type used in `libp2p-core`.

Perhaps it is too early to try and remove this abstraction. Here is a different idea:

- Resurrect https://github.com/libp2p/rust-libp2p/pull/3746
- Still introduce a `Protocol` newtype but have it in `libp2p-swarm`. This means protocols can use it and we can use it within our event types and commands